File: C++/transpose_matrix.cpp
#include <iostream>
#include "./transpose_matrix.h"
#include "./utility.h"

using namespace std;

vector<vector<int> > Solution::transpose(vector<vector<int> >& A) {
    vector<vector<int> > output;
    int row = A.size();
    int col = A[0].size();
    // we have to swap the row and column
    // O(col*row) -> O(N), where N is the total number of elements
    for (int i = 0; i < col; ++i) {
        vector<int> new_row;
        for (int j = 0; j < row; ++j) {
            new_row.push_back(A[j][i]);
        }
        output.push_back(new_row);
    }

    return output;
}

int main() {
    vector<vector<int> > A;
    
    vector<int> r1;
    r1.push_back(1);
    r1.push_back(2);
    r1.push_back(3);
    A.push_back(r1);
    vector<int> r2;
    r2.push_back(4);
    r2.push_back(5);
    r2.push_back(6);
    A.push_back(r2);
    

    Utility::printVector(A[0]);
    Utility::printVector(A[1]);
    
    Solution* sol = new Solution();
    
    vector<vector<int> > output = sol->transpose(A);
    delete sol;

    Utility::printVector(output[0]);
    Utility::printVector(output[1]);
    Utility::printVector(output[2]);

    return 0;
}

File: Python2/transpose_matrix.py
class Solution(object):
    def transpose(self, A):
        output = []
        row = len(A)
        col = len(A[0])
        for i in range(col):
            new_row = []
            for j in range(row):
                new_row.append(A[j][i])
            output.append(new_row)
        return output

    def unzip(self, A):
        return zip(*A)

    def list_comprehension(self, A):
        return [[row[col] for row in A] for col in range(len(A[0]))]


if __name__ == '__main__':
    A = [[1, 2, 3], [4, 5, 6]]
    sol = Solution()
    print sol.transpose(A)
    print sol.unzip(A)
    print sol.list_comprehension(A)
